lib/serializer/cbor/aws_iot_serializer_tinycbor_decoder.c

```diff
@@ -23,382 +23,453 @@
+                       pDecoderObject->value.stringLength = length;
+                   }
+
+                   break;
+               }
+
+            default:
+                /* Other scalar types are not supported. */
+                returnedError = AWS_IOT_SERIALIZER_NOT_SUPPORTED;
         }
     }
 
-    return returnError;
+    return returnedError;
 }
 
 /*-----------------------------------------------------------*/
 
-static AwsIotSerializerError_t _init(AwsIotSerializerDecoderObject_t * pDecoderObject,
-    uint8_t * pDataBuffer,
-    size_t maxSize)
+static AwsIotSerializerError_t _init( AwsIotSerializerDecoderObject_t * pDecoderObject,
+                                      uint8_t * pDataBuffer,
+                                      size_t maxSize )
 {
-    CborParser *pCborParser = pvPortMalloc(sizeof(CborValue));
-    _cborDecoder_t * pCborDecoder = pvPortMalloc(sizeof(_cborDecoder_t));
-    AwsIotSerializerError_t returnError = AWS_IOT_SERIALIZER_SUCCESS;
-    CborError cborError;
+    CborParser * pCborParser = pvPortMalloc( sizeof( CborParser ) );
+    _cborValueWrapper_t * pCborValueWrapper = pvPortMalloc( sizeof( _cborValueWrapper_t ) );
+
+    if( ( pCborParser == NULL ) || ( pCborValueWrapper == NULL ) )
+    {
+        vPortFree( pCborParser );
+        vPortFree( pCborValueWrapper );
+
+        return AWS_IOT_SERIALIZER_OUT_OF_MEMORY;
+    }
+
+    AwsIotSerializerError_t returnedError = AWS_IOT_SERIALIZER_SUCCESS;
+    CborError cborError = CborNoError;
 
     cborError = cbor_parser_init(
         pDataBuffer,
         maxSize,
         0,
         pCborParser,
-        &pCborDecoder->cborValue);
+        &pCborValueWrapper->cborValue );
 
-    if (_noCborError(cborError))
+    /* If init succeeds, create the decoder object. */
+    if( cborError == CborNoError )
     {
-        pCborDecoder->isOutermost = true;
-        _createDecoderObject(pCborDecoder, pDecoderObject);
+        pCborValueWrapper->isOutermost = true;
+
+        returnedError = _createDecoderObject( pCborValueWrapper, pDecoderObject );
     }
-    else
+
+    /* If there is any error or decoder object is a scalar type, free the cbor resources. */
+    if( cborError || returnedError || !_isArrayOrMap( pDecoderObject->type ) )
     {
+        /* pDecoderObject is untouched. */
+
         vPortFree( pCborParser );
-        vPortFree( pCborDecoder );
-        returnError = AWS_IOT_SERIALIZER_INTERNAL_FAILURE;
+        vPortFree( pCborValueWrapper );
     }
 
-    return returnError;
+    _translateErrorCode( cborError, &returnedError );
+
+    return returnedError;
 }
 
 /*-----------------------------------------------------------*/
 
-static AwsIotSerializerError_t _get(AwsIotSerializerDecoderIterator_t iterator,
-    AwsIotSerializerDecoderObject_t * pValueObject)
+static AwsIotSerializerError_t _destroy( AwsIotSerializerDecoderObject_t * pDecoderObject )
```

Comment:
As discussed lets change to void return type?

Reply:
yes